<a href="https://colab.research.google.com/github/ankur-srivastava/Applied-AI/blob/master/CNN_MNIST_Architecture.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
# Credits: https://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py

from __future__ import print_function
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K

batch_size = 128
num_classes = 10
epochs = 5

# input image dimensions
img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)


# This is the first architecture. A simple modification to the one provided by 
# Applied AI Team

# Using three CONV2D with kernel_size = (2, 2)
# Input -> CONV2D (2*2) -> CONV2D (2*2) -> MAX-POOL (2*2) -> CONV2D (2*2) 
# -> MAX_POOL(2*2) -> DROPOUT


modelA = Sequential()
modelA.add(Conv2D(32, kernel_size=(2, 2),
                 activation='relu',
                 input_shape=input_shape))
modelA.add(Conv2D(64, (2, 2), activation='relu'))
modelA.add(MaxPooling2D(pool_size=(2, 2)))
modelA.add(Conv2D(64, (5, 5), activation='relu'))
modelA.add(MaxPooling2D(pool_size=(2, 2)))
modelA.add(Dropout(0.25))
modelA.add(Flatten())
modelA.add(Dense(128, activation='relu'))
modelA.add(Dropout(0.5))
modelA.add(Dense(num_classes, activation='softmax'))

modelA.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

modelA.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = modelA.evaluate(x_test, y_test, verbose=0)
print('Test loss for model A:', score[0])
print('Test accuracy for model B:', score[1])


# This is the second architecture. Using a combination of 5*5 and 2*2

# Using two CONV2D with kernel_size = (5, 5)
# Input -> CONV2D (5*5) -> MAX-POOL (3*3) -> CONV2D (2*2) 
# -> MAX_POOL(2*2) -> DROPOUT

modelB = Sequential()
modelB.add(Conv2D(32, kernel_size=(5, 5),
                 activation='relu',
                 input_shape=input_shape))
modelB.add(MaxPooling2D(pool_size=(3, 3)))
modelB.add(Conv2D(64, (2, 2), activation='relu'))
modelB.add(MaxPooling2D(pool_size=(2, 2)))
modelB.add(Dropout(0.25))
modelB.add(Flatten())
modelB.add(Dense(128, activation='relu'))
modelB.add(Dropout(0.5))
modelB.add(Dense(num_classes, activation='softmax'))

modelB.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

modelB.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
scoreB = modelB.evaluate(x_test, y_test, verbose=0)
print('Test loss for model B:', scoreB[0])
print('Test accuracy for model B:', scoreB[1])

# This is the third architecture. Using a combination of 5*5 and 2*2

# Using three CONV2D
# Input -> CONV2D (2*2) -> MAX-POOL (2*2) -> CONV2D (3*3) 
# -> MAX_POOL(3*3) -> CONV2D (2*2) -> MAX-POOL (2*2) -> DROPOUT

modelC = Sequential()
modelC.add(Conv2D(32, kernel_size=(2, 2),
                 activation='relu',
                 input_shape=input_shape))
modelC.add(MaxPooling2D(pool_size=(2, 2)))
modelC.add(Conv2D(64, (3, 3), activation='relu'))
modelC.add(MaxPooling2D(pool_size=(3, 3)))
modelC.add(Conv2D(64, (2, 2), activation='relu'))
modelC.add(MaxPooling2D(pool_size=(2, 2)))
modelC.add(Dropout(0.25))
modelC.add(Flatten())
modelC.add(Dense(128, activation='relu'))
modelC.add(Dropout(0.5))
modelC.add(Dense(num_classes, activation='softmax'))

modelC.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

modelC.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
scoreC = modelC.evaluate(x_test, y_test, verbose=0)
print('Test loss for model C:', scoreC[0])
print('Test accuracy for model C:', scoreC[1])

x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/5
60000/60000 [==============================] - 209s 3ms/step - loss: 0.3170 - acc: 0.8995 - val_loss: 0.0633 - val_acc: 0.9788
Epoch 2/5
60000/60000 [==============================] - 208s 3ms/step - loss: 0.1009 - acc: 0.9697 - val_loss: 0.0387 - val_acc: 0.9884
Epoch 3/5
60000/60000 [==============================] - 208s 3ms/step - loss: 0.0721 - acc: 0.9789 - val_loss: 0.0307 - val_acc: 0.9897
Epoch 4/5
60000/60000 [==============================] - 208s 3ms/step - loss: 0.0569 - acc: 0.9830 - val_loss: 0.0252 - val_acc: 0.9911
Epoch 5/5
60000/60000 [==============================] - 211s 4ms/step - loss: 0.0506 - acc: 0.9847 - val_loss: 0.0231 - val_acc: 0.9924
Test loss for model A: 0.023069058604503517
Test accuracy for model B: 0.9924
Train on 60000 samples, validate on 10000 samples
Epoch 1/5
60000/60000 [==============================] - 47s 776

**From earlier outputs for the three Models we can conclude the following -**

For Model A with 3 convolution layers the accuracy is 99.24%

For Model B with 2 convolution layers the accuracy is 98.96%

For Model C with 2 convolution layers the accuracy is 96.71%